### ISSUE TITLE
Ghost atom masses and molecular weight

### ISSUE THREAD

**Issue by bgpeyton** [psi4/psi4#909]
This is an issue I came across when needing to calculate the molecular weight of a molecule (which I was surprised is not already part of the Molecule class). 

Currently, ghost atoms still return a mass when calling `mol.mass(i)`. The best way I have found to avoid this on the Python side is to check the Z value `if mol.Z():` because ghost atoms (correctly) return false here. `optrot.cc` does not do this correctly and so uses an incorrect molecular weight for molecules with ghost atoms, and there may be more codes that have this issue. I thought it may be better to fix `mass()` rather than try to find every single code which calculates a molecular weight and add in the Z value check.

I feel ghost atoms should not have any mass in the first place, as they are usually just coordinates on which to place basis functions corresponding to the atom. I can't think of any places where you would want the mass of a ghosted atom, so please correct me if I'm mistaken. 

Within `masses.h`, a note is found:
```
** 10/08/99 EFV - Changed an2masses[0] to 0.0000
**                (ghost atom's weight should be 0)
```
But in `molecule.cc` the an2masses list is only used if `atoms_[atom]->mass() != 0` returns false, which will not happen for ghost atoms as they are currently done. Instead, it should probably be checked first `if atoms_[atom]->Z() == 0` and return a zero mass. (I thought this was better than pushing the Z == 0 case into the else statement, because doing that throws a _lot_ of warnings at the user). This is an easy fix, but as this effects the molecule class I wanted to check some other opinions before opening my PR. There also may be a more fundamental place to fix this problem, such as when the `atoms_` list is generated, but I wasn't sure.

**Comment by loriab** [comment]
Your analysis of inconsistency and proposed fix are totally reasonable. But I disagree and rather have the opposite complaint about the Molecule class – that Z returns 0 instead of atomic number for ghost atoms so that function is actually conflating two independent quantities. mass is doing the right thing.

The reason nonzero quantities are stored for ghost atoms in a CoordEntry is because the atoms can change their ghostedness upon `extract_subsets`, `activate_all_fragments`, etc. This leads to two tiers of ghostedness. The first are those that are read in as ghosts `@C` and get 0 Z and charge from the start. The others are those that are read in as real, then have their ghostedness changed on the fly. Note that Molecule internal fns sometimes have to check both `Molecule.fragment_types_` and `CoordEntry.ghost_` to get the full state of the Molecule.

I've taken the back-away-slowly approach to ghosts in Molecule. And I have some memory of full-real-mol --> ghost-fragment-mol --> full-real-mol not actually being able to accomplish that last step. We tend to clone-manipulate-discard Mols these days rather than transform them internally, so our tests may not have noticed.

In the [molrec dictionary repr](https://github.com/loriab/psi4/blob/pyvib2/psi4/driver/qcdb/molparse/tofrom.py#L78-L113) of a molecule, I've dropped fragment_types and consolidated the ghostedness. I've tried to keep the fields either totally independent (Z from real) or tied together (Z and symbol; to a lesser extent A, Z, symbol, mass).

As far as getting things working for you, i'm inclined to suggest letting sleeping ghosts lie as far as internal Molecule structure. But go ahead and write a `molecular_weight` fn that accounts for both types of ghostedness internally and has a required bool argument as to whether or not mass of ghosts should be accounted for.

Other opinions or rebuttals welcome.

**Comment by bgpeyton** [comment]
You're right, the inconsistency could be fixed the other way around by "fixing" the Z values instead. Though even if a new `molecular_weight` fn were implemented in `molecule` that lets me circumvent the mass issue when necessary, the problem is that there may be places other than `optrot.cc` that are returning incorrect masses without even knowing it. I suppose these codes could be tracked down and updated with some hunting. I'd be more than happy to make a PR and add such a function, and keep my eyes out for codes calculating molecular weight.

However, I'm still not sure I understand the utility of retaining masses for ghosted atoms. With some testing, it looks like real -> ghost -> real is still possible using `extract_subsets` even if I set the mass of ghost atoms to zero. What doesn't seem to be possible, on the other hand, is ghost -> real, but that doesn't seem to work right now anyway (not really sure why that is, may look into this if I have time). I tested this just by editing `Molecule::mass` in the way described above.

I'm very unfamiliar with how masses and fragment identities are passed around. Is there a case in which having the ghost masses zero would keep you from making a properly-massed all-real-atoms copy of that molecule? That would definitely preclude any option of setting them zero permanently.

**Comment by amjames** [comment]
I have no horse in this race in terms of is the behavior of `Molecule.mass` or `Molecule.Z` acting strangely, but the inconstancy is a problem.

If we took every property test in the suite and added a test where ghost atoms are involved I think we would turn up some failures since the way molecular weight is calculated in `optrot.cc` as @bgpeyton described does not properly ignore ghosted atoms and i wouldn't be surprised if there were others.

**Comment by loriab** [comment]
I completely agree with @bgpeyton and @amjames that there are probably misuses of Z & mass in the code resulting from unclear or inconsistent definitions (I think there's even cases where the `Mol.property()` fn differs in behavior from the `Mol.fproperty()` fn (that includes dummy atoms). It would be great to survey the code and hunt these down.

My trepidation comes in because Mol is delicate w/the two versions of ghosting (type A (not reversible): `@C`, type B (reversible): `extract_subset(1, 2)`) and I worry that changes would start an avalanche. In fact I definitely recommend _not_ "fixing" `Mol.Z()` the way I would like (`C` and `@C` both return 6) because lots of places in the Mol code depend on using that fn as a test for type A ghostedness.

> I'm very unfamiliar with how masses and fragment identities are passed around. Is there a case in which having the ghost masses zero would keep you from making a properly-massed all-real-atoms copy of that molecule? That would definitely preclude any option of setting them zero permanently.

What happens is that every atom at parse-time gets read into a CoordEntry object that holds Z, ghost, charge, mass, element symbol, coords, etc. If the atom is type-A ghost, then Z, ghost, and charge get set accordingly. There is no other info so type-A is not eligible to return to reality. The fragmentation pattern is also stored at parse-time with all-`Real`. For a trimer, when you `extract_subsets(1, 2)`, the `Mol.fragment_types_` gets reset to `[Real, Ghost, Absent]`. The CoordEntries don't change, but every function in Mol is trained to iterate over the frags, and handle Ghost or Absent differently, so that in the fn output, they _appear_ ghosted (type B) or absent. Because the mass, Z info in the CoordEntry is unchanged, this ghostedness is reversible. So in the type B sense, mass must not be set to 0 permanently b/c then type B would be irreversible. I suppose you're advocating zeroing mass for type A? That'd be more consistent, but its effects on COM, inertial frame, symmetry, etc. would need to be probed. Or maybe our moments of inertia are even now wrong in the presence of ghosts.

Summary: I would love a rationalization pass through the Mol accessor fns. But what I value more than that is (1) Mol not breaking for downstream code we don't control, (2) a pass through the psi codebase making sure `Z` and `mass` are being used in accordance with the programmers' intentions (apparently not in optrot), and (3) you not going down a Mol rabbit hole.

Maybe we should talk more at the conf call tomorrow.

**Comment by loriab** [comment]
Note that I have no doubt Mol was thoroughly tested for all these things in its development and was in pristine working order. But we didn't have the notion of preserving those unit tests for a future refactoring. So I'm just trying to convey that downstream may rely on the inconsistency and changes that in class development may have been a "fix" are now a "refactoring".

**Comment by loriab** [comment]
One reason ghosted masses may be nonzero is we didn't want the frame to change btwn a fully real dimer and its counterpoise variants? wouldn't be surprised if sapt relies on this. That brings to mind a question I had on the Hessian analysis of whether frame should be changing upon isotopic substitution. Center of charge frame would solve that but not a frame-change-upon-ghosting issue.

Upshot is that some uses require ghost Z be zero (nuclear repulsion energy), some require ghost Z be atomic number (reading off element identity), some require ghost mass be zero (molecular mass), some require ghost mass be real (frame stability for dimer subsetting). So may as well provide a Mol function with a required argument to choose each case (like ACS & DGAS recently imposed for nuclear_repulsion_energy) to make sure consumers are aware of the differences and are using the right one. Python export can defuault to something sensible (hopefully the current behavior).

**Comment by bgpeyton** [comment]
I see that this goes a lot further than I was expecting it to; some day that will stop surprising me. It's a shame about the apparent necessity to keep the inconsistent definitions, but you're probably right when you say lots of downstream code is going to rely on the current behavior, whether or not we see it as "correct." The dual-identity of type-A and type-B ghosts only complicates matters further, though I definitely see the need for it. It would almost be better if type-A ghosts didn't exist and at parse-time any "default" ghosts are treated as type-B ghosts somehow, but this is likely more complicated then it sounds. Either way, the ability to go back-and-forth between real and ghost is absolutely necessary, and type-B is the only one that can consistently do that. 

Fixing only type-A ghosts to return zero mass may only confuse things further. If nothing else, the functions should at least do the same thing for type-A and type-B ghosts. Consistency between `Molecule.mass` and `Molecule.Z` could be worked on later, though as you pointed out it's probably not a good idea to change `Molecule.Z`. 

As for a new `Molecule` function to ensure the user knows what they're going to get, are you referring specifically to a `molecular_weight` function or more generally a function to return either the mass or Z val as either zero or real depending on input?  I could get behind either of those, though with `molecular_weight` I would suggest simply setting a default for ghost atoms to return zero mass as I can't imagine anyone would want a molecular weight including ghost atom masses.

**Comment by tsexton** [comment]
If you are doing an update of the documentation for Molecule.Z, it might be useful to consider possible user confusion between Molecule.Z (capital, gives the charge) and Molecule.z (lowercase, gives the z-coordinate).  Both return floats, so no type mismatch, and it's not an obvious typo visually.  Maybe a warning in the API would be helpful.

**Comment by loriab** [comment]
> It would almost be better if type-A ghosts didn't exist and at parse-time any "default" ghosts are treated as type-B ghosts somehow

That's what I do for the json molecule record – there's a list `real = [True, False, True, False, False, False]` and a separate `fragment_separators = [3]` that decouples what would be the following in psi. Ghosting types A & B are collapsed. 

```
molecule {
C
@C
C
--
C
C
C
}
extract_subsets(1, 2)
```

> As for a new Molecule function to ensure the user knows what they're going to get, are you referring specifically to a molecular_weight function or more generally a function to return either the mass or Z val as either zero or real depending on input?

I think I've settled on recommending changing the signature of `Molecule.Z()` and `Molecule.mass()` to take a required argument `bool zero_ghost`. The recompile will force consideration of which was wanted everywhere (incl. downstream). At the same time, I think leave `CoordEntry.Z()` and `CoordEntry.mass()` alone in their probably inconsistent state. Most of the `Z()` calls in Mol are probably to `CoordEntry.Z()` and have the gh=0 factored into the logic.

Should definitely get consensus tomorrow before diving in.

**Comment by loriab** [comment]
@tsexton, yes, it's a pity that the name `Z` which should properly be `int` atomic number has devolved into `float` nuclear charge.

I support all documentation clarifications. You can even edit [them](https://github.com/psi4/psi4/blob/master/psi4/src/export_mints.cc#L985-L993) online with the little pencil. GH will turn your edits into a pull request.
